cumulus_library/statistics/psm.py

```diff
@@ -0,0 +1,247 @@
+        cursor.execute(drop)
+        dataset_query = get_create_covariate_table(
+            target_table=self.config.target_table,
+            pos_source_table=self.config.pos_source_table,
+            neg_source_table=self.config.neg_source_table,
+            primary_ref=self.config.primary_ref,
+            dependent_variable=self.config.dependent_variable,
+            join_cols_by_table=self.config.join_cols_by_table,
+            count_ref=self.config.count_ref,
+            count_table=self.config.count_table,
+        )
+        self.queries.append(dataset_query)
+
+    def generate_psm_analysis(self, cursor: object, schema: str):
+        """Runs PSM statistics on generated tables"""
+        df = cursor.execute(f"select * from {self.config.target_table}").as_pandas()
+        symptoms_dict = self._get_symptoms_dict(self.config.classification_json)
+        for dependent_variable, codes in symptoms_dict.items():
+            df[dependent_variable] = df["code"].apply(lambda x: 1 if x in codes else 0)
+        df = df.drop(columns="code")
+        # instance_count present but unused for PSM if table contains a count_ref input
+        # (it's intended for manual review)
+        df = df.drop(columns="instance_count", errors="ignore")
+
+        columns = []
+        if self.config.join_cols_by_table is not None:
+            for table_key in self.config.join_cols_by_table:
+                for column in self.config.join_cols_by_table[table_key][
+                    "included_cols"
+                ]:
+                    if len(column) == 2:
+                        columns.append(column[1])
+                    else:
+                        columns.append(column[0])
+
+        for column in columns:
+            encoded_df = pandas.get_dummies(df[column])
+            df = pandas.concat([df, encoded_df], axis=1)
+            df = df.drop(column, axis=1)
+        df = df.reset_index()
+
+        try:
+            psm = PsmPy(
+                df,
+                treatment=self.config.dependent_variable,
+                indx=self.config.primary_ref,
+                exclude=[],
+            )
+            # This function populates the psm.predicted_data element, which is required
+            # for things like the knn_matched() function call
+            psm.logistic_ps(balance=True)
+            print(psm.predicted_data)
+            # This function populates the psm.df_matched element
+            # TODO: flip replacement to false after increasing sample data size
+            psm.knn_matched(
+                matcher="propensity_logit",
+                replacement=True,
+                caliper=None,
+                drop_unmatched=True,
+            )
+            print(psm.df_matched)
+        except ZeroDivisionError:
+            print(
+                "Encountered a divide by zero error during statistical graph generation. Try increasing your sample size."
+            )
```

Comment:
Just a print and not an exit, for these two exceptions? (nit: ideally would pipe to stderr instead of stdout)

Reply:
or, jesus, raise an error or something reasonable